apps/src/templates/instructions/TopInstructions.jsx

```diff
@@ -619,222 +559,115 @@ class TopInstructions extends Component {
-                    teacherOnly={teacherOnly}
-                    isMinecraft={this.props.isMinecraft}
-                    isRtl={this.props.isRtl}
-                  />
-                )}
-              {this.props.viewAs === ViewType.Teacher &&
-                (this.props.teacherMarkdown || showContainedLevelAnswer) && (
-                  <InstructionsTab
-                    className="uitest-teacherOnlyTab"
-                    onClick={this.handleTeacherOnlyTabClick}
-                    selected={this.state.tabSelected === TabType.TEACHER_ONLY}
-                    text={i18n.teacherOnly()}
-                    teacherOnly={teacherOnly}
-                    isMinecraft={this.props.isMinecraft}
-                    isRtl={this.props.isRtl}
-                  />
-                )}
-            </div>
-            {/* For CSF contained levels we use the same collapse function as CSD/CSP*/}
-            {!this.props.isEmbedView &&
-              (isCSDorCSP || this.props.hasContainedLevels) && (
-                <CollapserIcon
-                  isCollapsed={this.props.isCollapsed}
-                  onClick={this.handleClickCollapser}
-                  style={collapserIconStyles}
-                />
-              )}
-          </div>
-        </PaneHeader>
-        <div
-          style={[this.props.isCollapsed && isCSDorCSP && commonStyles.hidden]}
-        >
+          tabSelected={tabSelected}
+          isCSDorCSP={isCSDorCSP}
+          displayHelpTab={displayHelpTab}
+          displayFeedback={displayFeedback}
+          displayKeyConcept={displayKeyConcept}
+          isViewingAsTeacher={this.isViewingAsTeacher()}
+          fetchingData={fetchingData}
+          handleDocumentationClick={this.handleDocumentationClick}
+          handleInstructionTabClick={this.handleInstructionTabClick}
+          handleHelpTabClick={this.handleHelpTabClick}
+          handleCommentTabClick={this.handleCommentTabClick}
+          handleTeacherOnlyTabClick={this.handleTeacherOnlyTabClick}
+          handleClickCollapser={this.handleClickCollapser}
+        />
+        <div style={[isCollapsed && isCSDorCSP && commonStyles.hidden]}>
           <div
             style={[
               isCSF &&
-              !this.props.hasContainedLevels &&
-              this.state.tabSelected === TabType.INSTRUCTIONS
+              !hasContainedLevels &&
+              tabSelected === TabType.INSTRUCTIONS
                 ? styles.csfBody
                 : styles.body,
-              this.props.isMinecraft && craftStyles.instructionsBody
+              isMinecraft && craftStyles.instructionsBody
             ]}
             id="scroll-container"
           >
-            <div
-              ref={ref => {
-                if (ref) {
-                  this.instructions = ref;
-                }
-              }}
-            >
-              {this.props.hasContainedLevels && (
+            <div ref={ref => this.setInstructionsRef(ref)}>
```

Comment:
Maybe create a renderInstructions() function that returns one of the three possible cases of instructions?